controllers/openstackmachine_controller.go

```diff
@@ -431,14 +440,24 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 }
 
 func (r *OpenStackMachineReconciler) getOrCreate(logger logr.Logger, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster, machine *clusterv1.Machine, openStackMachine *infrav1.OpenStackMachine, computeService *compute.Service, userData string) (*compute.InstanceStatus, error) {
-	instanceStatus, err := computeService.GetInstanceStatusByName(openStackMachine, openStackMachine.Name)
-	if err != nil {
-		logger.Info("Unable to get OpenStack instance", "name", openStackMachine.Name)
-		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.OpenStackErrorReason, clusterv1.ConditionSeverityError, err.Error())
-		return nil, err
+	var instanceStatus *compute.InstanceStatus
+	var err error
+	if openStackMachine.Spec.InstanceID != nil {
+		instanceStatus, err = computeService.GetInstanceStatus(*openStackMachine.Spec.InstanceID)
+		if err != nil {
+			logger.Info("Unable to get OpenStack instance", "name", openStackMachine.Name)
+			conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.OpenStackErrorReason, clusterv1.ConditionSeverityError, err.Error())
+			return nil, err
+		}
 	}
 
 	if instanceStatus == nil {
+		// First check if there is an existing instance with machine name, in case where instance ID would not have been stored in machine status
+		if instanceStatus, err = computeService.GetInstanceStatusByName(openStackMachine, openStackMachine.Name); err == nil {
+			if instanceStatus != nil {
```

Comment:
We should set `openStackMachine.Spec.InstanceID` here.

Reply:
For now `openStackMachine.Spec.InstanceID` was set right after `getOrCreate` [function call](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/f271006c0dd0cc60df682f60ed969503e07eb0c2/controllers/openstackmachine_controller.go#L346C11-L346C11). I'll do it here instead, but it also needs to be done before final return of `getOrCreate` in that case.

Reply:
Yes, on the other hand, setting InstanceID here could also help ensuring that it will always be done as soon as we retrieve it, whereas if we do it in `reconcileNormal` fiunction, it may be delayed by some future refactoring that would add code between the call to `getOrCreate` and the setting of the ID....

I've switched back to previous version as it's slighlty cleaner as you said.